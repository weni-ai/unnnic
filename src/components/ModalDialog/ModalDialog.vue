<template>
  <section
    v-if="modelValue"
    class="unnnic-modal-dialog"
  >
    <section
      class="unnnic-modal-dialog__overlay"
      @click.stop="!persistent && close()"
    />
    <section
      :class="[
        'unnnic-modal-dialog__container',
        `unnnic-modal-dialog__container--${size}`,
      ]"
    >
      <header
        v-if="title"
        class="unnnic-modal-dialog__container__header"
      >
        <section class="unnnic-modal-dialog__container__title-container">
          <UnnnicIcon
            v-if="icon || type"
            class="unnnic-modal-dialog__container__title-icon"
            :icon="icon || iconsMapper[type].icon"
            :scheme="iconScheme || iconsMapper[type].scheme"
            size="md"
          />
          <h1 class="unnnic-modal-dialog__container__title-text">
            {{ title }}
          </h1>
        </section>
        <UnnnicIcon
          v-if="showCloseIcon"
          icon="close"
          clickable
          @click="close()"
        />
      </header>
      <section class="unnnic-modal-dialog__container__content">
        <slot></slot>
      </section>
      <section
        v-if="primaryButtonText"
        :class="[
          'unnnic-modal-dialog__container__actions',
          showActionsDivider &&
            'unnnic-modal-dialog__container__actions--divider',
        ]"
      >
        <UnnnicButton
          type="tertiary"
          :text="secondaryButtonText || i18n('cancel')"
          @click.stop="
            secondaryButtonText ? $emit('secondaryButtonClick') : close()
          "
        />
        <UnnnicButton
          :type="
            primaryButtonType || primaryButtonTypeMapper[type] || 'primary'
          "
          :text="primaryButtonText"
          @click.stop="$emit('primaryButtonClick')"
        />
      </section>
    </section>
  </section>
</template>

<script>
import UnnnicIcon from '../Icon.vue';
import UnnnicButton from '../Button/Button.vue';
import UnnnicI18n from '../../mixins/i18n';

export default {
  name: 'UnnnicModalDialog',
  components: {
    UnnnicIcon,
    UnnnicButton,
  },
  mixins: [UnnnicI18n],
  props: {
    modelValue: {
      type: Boolean,
      required: true,
    },
    persistent: {
      type: Boolean,
      default: false,
    },
    type: {
      type: String,
      default: '',
      validate(type) {
        return ['success', 'warning', 'attention'].includes(type);
      },
    },
    size: {
      type: String,
      default: 'md',
      validate(size) {
        return ['sm', 'md', 'lg'].includes(size);
      },
    },
    title: {
      type: String,
      default: 'title',
    },
    icon: {
      type: String,
      default: '',
    },
    iconScheme: {
      type: String,
      default: '',
    },
    showCloseIcon: {
      type: Boolean,
      default: false,
    },
    primaryButtonText: {
      type: String,
      default: '',
    },
    secondaryButtonText: {
      type: String,
      default: '',
    },
    primaryButtonType: {
      type: String,
      default: '',
    },
    showActionsDivider: {
      type: Boolean,
      default: false,
    },
  },
  emits: ['primaryButtonClick', 'secondaryButtonClick', 'update:modelValue'],

  data() {
    return {
      defaultTranslations: {
        cancel: {
          'pt-br': 'Cancelar',
          en: 'Cancel',
          es: 'Cancelar',
        },
      },
      iconsMapper: {
        success: { icon: 'check_circle', scheme: 'aux-green-500' },
        warning: { icon: 'warning', scheme: 'aux-red-500' },
        attention: { icon: 'error', scheme: 'aux-yellow-500' },
      },
      primaryButtonTypeMapper: {
        success: 'primary',
        warning: 'warning',
        attention: 'attention',
      },
    };
  },
  methods: {
    close() {
      this.$emit('update:modelValue', false);
    },
  },
};
</script>

<style lang="scss" scoped>
@import '../../assets/scss/unnnic.scss';
* {
  margin: 0;
  padding: 0;
  box-sizing: border-box;
}
.unnnic-modal-dialog {
  width: 100vw;
  height: 100vh;
  position: fixed;
  left: 0;
  top: 0;
  display: flex;
  justify-content: center;
  align-items: center;
  z-index: 9999;

  &__overlay {
    position: absolute;
    width: 100%;
    height: 100%;
    background-color: rgba(0, 0, 0, 0.4);
  }
}

.unnnic-modal-dialog__container {
  background: $unnnic-color-neutral-white;
  border-radius: $unnnic-spacing-xs;
  box-shadow: 0px 4px 8px 0px rgba(0, 0, 0, 0.1);
  position: fixed;

  &--sm {
    width: 400px;
  }
  &--md {
    width: 600px;
  }
  &--lg {
    width: 800px;
  }

  &__header {
    display: flex;
    justify-content: space-between;
    align-items: center;
    border-bottom: 1px solid $unnnic-color-neutral-soft;
    padding: $unnnic-spacing-md;
  }

  &__title-container {
    display: flex;
    align-items: center;
    gap: $unnnic-spacing-ant;
  }

  &__title-icon {
    font-size: 28px;
  }

  &__title-text {
    font-family: $unnnic-font-family-secondary;
    font-size: $unnnic-font-size-title-sm;
    font-weight: $unnnic-font-weight-black;
    line-height: 28px;
    color: $unnnic-color-neutral-darkest;
  }

  &__content {
    padding: $unnnic-spacing-md;
    color: $unnnic-color-neutral-cloudy;
  }

  &__actions {
    display: flex;
    gap: 16px;
    padding: $unnnic-spacing-md;
    > * {
      flex-grow: 1;
    }
    &--divider {
      border-top: 1px solid $unnnic-color-neutral-soft;
    }
  }
}
</style>
