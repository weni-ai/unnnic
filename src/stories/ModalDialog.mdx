import { Meta, Source, Story } from '@storybook/blocks';

import * as ModalDialogStories from './ModalDialog.stories';

<Meta of={ModalDialogStories}/>

# ModalDialog

The ModalDialog was designed to inform the user about a specific task, request information, or involve other tasks.

<Source
  language='html'
  dark
  code={`
    <UnnnicModalDialog v-model="showModal">
    <p>Modal Content</p>
</UnnnicModalDialog>
  `}
/>

---

#### **Props Options:**

| Key                 | Description                                                                                                                                   | Values                                | Default |
|---------------------|-----------------------------------------------------------------------------------------------------------------------------------------------|---------------------------------------|---------|
| type                | Modal type predefinition                                                                                                                      | 'success' \| 'warning' \| 'attention' | ''      |
| persistent          | Clicking outside of the element will not deactive it                                                                                          | true \| false                         | false   |
| size                | Modal width size (sm = 400px; md = 600px; lg = 800px)                                                                                         | 'sm' \| 'md' \| 'lg'                  | 'md'    |
| title               | Modal card title (if not provided, the title section will not be displayed)                                                                   | string                                | ''      |
| icon                | Custom title section icon (if not provided, the preset icons will be considered according to the chosen modal type)                           | string                                | ''      |
| iconScheme          | Custom scheme color to icon in title section (if not provided, the preset scheme color will be considered according to the chosen modal type) | string                                | ''      |
| showCloseIcon       | Show close icon in title section                                                                                                              | boolean                               | false   |
| primaryButtonText   | Button primary action text (if not provided, the actions section will not be displayed)                                                       | string                                | ''      |
| secondaryButtonText | Button secondary action text (if not provided and there is primaryButtonText, this button will take 'Cancel' action                           | string                                | ''      |
| primaryButtonType   | Button primary type (if not provided, defaults to the specified type)                                                                         | string                                | ''      |
| showActionsDivider  | Show divider in between actions and dialog content                                                                                            | boolean                               | false   |

## Examples
Some examples of uses of the modal

### Default

<Story of={ModalDialogStories.Default} />

### Warning

<Story of={ModalDialogStories.Warning} />

### Attention

<Story of={ModalDialogStories.Attention} />

### Overflowed Content

<Story of={ModalDialogStories.Overflowed} />

### Form Content

<Story of={ModalDialogStories.Form} />

### Image Content

<Story of={ModalDialogStories.Image} />