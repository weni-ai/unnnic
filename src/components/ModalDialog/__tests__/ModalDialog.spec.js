import { mount } from '@vue/test-utils';
import { describe, it, vi } from 'vitest';
import ModalDialog from '../ModalDialog.vue';

describe('ModalDialog.vue', () => {
  let wrapper;

  beforeEach(() => {
    wrapper = mount(ModalDialog, {
      props: {
        title: 'Test Title',
        icon: 'test-icon',
        modelValue: true,
        showCloseIcon: true,
        primaryButtonProps: { text: 'Confirm' },
        'onUpdate:modelValue': (e) => wrapper.setProps({ modelValue: e }),
      },
      global: {
        stubs: ['UnnnicIcon', 'UnnnicButton'],
      },
    });
  });

  describe('Elements rendering', () => {
    it('matches the snapshot', () => {
      expect(wrapper.html()).toMatchSnapshot();
    });

    it('should render correctly when modelValue is true', () => {
      const modal = wrapper.find('[data-testid="modal-dialog"]');
      expect(modal.exists()).toBe(true);
    });

    it('should not render when modelValue is false', async () => {
      await wrapper.setProps({ modelValue: false });

      const modal = wrapper.find('[data-testid="modal-dialog"]');
      expect(modal.exists()).toBe(false);
    });

    it('should apply the correct size class based on the size prop', async () => {
      const modalContainer = wrapper.find('[data-testid="modal-container"]');
      const defaultClass = 'unnnic-modal-dialog__container';

      expect(modalContainer.classes()).toContain(defaultClass + '--md');

      await wrapper.setProps({ size: 'lg' });
      expect(modalContainer.classes()).toContain(defaultClass + '--lg');

      await wrapper.setProps({ size: 'sm' });
      expect(modalContainer.classes()).toContain(defaultClass + '--sm');
    });

    it('should render the icon and title when provided', () => {
      const title = wrapper.find('[data-testid="title-text"]');
      expect(title.exists()).toBe(true);
      expect(title.text()).toBe('Test Title');

      const icon = wrapper.findComponent('[data-testid="title-icon"]');
      expect(icon.exists()).toBe(true);
      expect(icon.props('icon')).toBe('test-icon');
    });

    it('should render the icon from iconsMapper when type prop is provided', async () => {
      await wrapper.setProps({
        type: 'success',
        icon: '',
      });

      const icon = wrapper.findComponent('[data-testid="title-icon"]');
      expect(icon.exists()).toBe(true);
      expect(icon.props().icon).toBe(wrapper.vm.iconsMapper['success'].icon);
    });

    it('should not render the icon when both icon and type props are not provided', async () => {
      await wrapper.setProps({
        icon: '',
        type: '',
      });

      const icon = wrapper.find('[data-testid="title-icon"]');
      expect(icon.exists()).toBe(false);
    });

    it('should render the close icon when showCloseIcon is true', () => {
      const closeIcon = wrapper.find('[data-testid="close-icon"]');
      expect(closeIcon.exists()).toBe(true);
    });

    it('should not render buttons when primaryButtonProps is not provided', async () => {
      await wrapper.setProps({ primaryButtonProps: undefined });

      const primaryButton = wrapper.find('[data-testid="primary-button"]');
      const secondaryButton = wrapper.find('[data-testid="secondary-button"]');
      expect(primaryButton.exists()).toBe(false);
      expect(secondaryButton.exists()).toBe(false);
    });
  });

  describe('Overlay behavior', () => {
    it('should close the modal when clicking on the overlay if persistent is false', async () => {
      expect(wrapper.props().modelValue).toBe(true);

      const overlay = wrapper.find('[data-testid="modal-overlay"]');
      await overlay.trigger('click');

      expect(wrapper.emitted('update:modelValue')).toBeTruthy();
      expect(wrapper.emitted('update:modelValue')[0]).toEqual([false]);
      expect(wrapper.props().modelValue).toBe(false);
    });

    it('should not close the modal when clicking on the overlay if persistent is true', async () => {
      await wrapper.setProps({ persistent: true });

      const overlay = wrapper.find('[data-testid="modal-overlay"]');
      await overlay.trigger('click');

      expect(wrapper.emitted('update:modelValue')).toBeFalsy();
      expect(wrapper.props().modelValue).toBe(true);
    });
  });

  describe('Slot rendering', () => {
    it('should render leftSidebar slot when provided', () => {
      const wrapper = mount(ModalDialog, {
        props: {
          modelValue: true,
        },
        slots: {
          leftSidebar:
            '<div data-testid="left-sidebar">Left Sidebar Content</div>',
        },
      });

      const leftSidebar = wrapper.find('[data-testid="left-sidebar"]');
      expect(leftSidebar.exists()).toBe(true);
      expect(leftSidebar.text()).toBe('Left Sidebar Content');
    });

    it('should render default slot content', () => {
      const wrapper = mount(ModalDialog, {
        props: {
          modelValue: true,
        },
        slots: {
          default: '<div data-testid="default-slot">Default Slot Content</div>',
        },
      });

      const defaultSlot = wrapper.find('[data-testid="default-slot"]');
      expect(defaultSlot.exists()).toBe(true);
      expect(defaultSlot.text()).toBe('Default Slot Content');
    });
  });

  describe('Buttons actions', () => {
    it('should close the modal when clicking on close icon', async () => {
      const closeIcon = wrapper.find('[data-testid="close-icon"]');
      await closeIcon.trigger('click');

      expect(wrapper.emitted('update:modelValue')).toBeTruthy();
      expect(wrapper.emitted('update:modelValue')[0]).toEqual([false]);
      expect(wrapper.props().modelValue).toBe(false);
    });

    it('should emit primaryButtonClick event when the primary button is clicked', async () => {
      const primaryButton = wrapper.find('[data-testid="primary-button"]');
      await primaryButton.trigger('click');

      expect(wrapper.emitted('primaryButtonClick')).toBeTruthy();
    });

    it('should emit secondaryButtonClick event when the secondary button is clicked', async () => {
      await wrapper.setProps({
        secondaryButtonProps: { text: 'Cancel' },
      });

      const secondaryButton = wrapper.find('[data-testid="secondary-button"]');
      await secondaryButton.trigger('click');

      expect(wrapper.emitted('secondaryButtonClick')).toBeTruthy();
    });

    it('should close the modal when the secondary button is clicked and no secondaryButtonClick event is provided', async () => {
      const secondaryButton = wrapper.find('[data-testid="secondary-button"]');
      await secondaryButton.trigger('click');

      const emittedValue = wrapper.emitted('update:modelValue');

      expect(emittedValue).toBeTruthy();
      expect(emittedValue[0][0]).toEqual(false);
    });
  });

  describe('Actions and appearance', () => {
    it('should not render the secondary button when hideSecondaryButton is true', async () => {
      await wrapper.setProps({
        hideSecondaryButton: true,
      });
      const secondaryButton = wrapper.find('[data-testid="secondary-button"]');
      expect(secondaryButton.exists()).toBe(false);
    });

    it('should apply a divider class to the actions section when showActionsDivider is true', async () => {
      await wrapper.setProps({
        showActionsDivider: true,
      });

      const actionsSection = wrapper.find('[data-testid="actions-section"]');
      expect(actionsSection.classes()).toContain(
        'unnnic-modal-dialog__container__actions--divider',
      );
    });
  });

  describe('Body overflow', () => {
    it('should toggle body overflow based on modal visibility', async () => {
      await wrapper.setProps({ modelValue: false });
      const updateBodyOverflowSpy = vi.spyOn(wrapper.vm, 'updateBodyOverflow');

      await wrapper.setProps({ modelValue: true });
      expect(updateBodyOverflowSpy).toHaveBeenCalledWith(true);

      await wrapper.setProps({ modelValue: false });
      expect(updateBodyOverflowSpy).toHaveBeenCalledWith(false);
    });
  });

  describe('Validators', () => {
    describe('type prop validator', () => {
      it('should validate type prop when a valid value is provided', () => {
        const validTypes = ['success', 'warning', 'attention'];
        validTypes.forEach((validType) => {
          expect(ModalDialog.props.type.validate(validType)).toBe(true);
        });
      });

      it('should invalidate type prop when an invalid value is provided', () => {
        expect(ModalDialog.props.type.validate('invalidType')).toBe(false);
      });
    });

    describe('size prop validator', () => {
      it('should validate size prop when a valid value is provided', () => {
        const validSizes = ['sm', 'md', 'lg'];
        validSizes.forEach((validSize) => {
          expect(ModalDialog.props.size.validate(validSize)).toBe(true);
        });
      });

      it('should invalidate size prop when an invalid value is provided', () => {
        expect(ModalDialog.props.size.validate('invalidSize')).toBe(false);
      });
    });
  });
});
