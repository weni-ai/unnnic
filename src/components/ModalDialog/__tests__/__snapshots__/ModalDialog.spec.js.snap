// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ModalDialog.vue > Elements rendering > matches the snapshot 1`] = `
"<section data-v-68ebadeb="" class="unnnic-modal-dialog" data-testid="modal-dialog">
  <section data-v-68ebadeb="" class="unnnic-modal-dialog__overlay" data-testid="modal-overlay"></section>
  <section data-v-68ebadeb="" class="unnnic-modal-dialog__container unnnic-modal-dialog__container--md" data-testid="modal-container">
    <!--v-if-->
    <section data-v-68ebadeb="" class="unnnic-modal-dialog__container__body">
      <header data-v-68ebadeb="" class="unnnic-modal-dialog__container__header">
        <section data-v-68ebadeb="" class="unnnic-modal-dialog__container__title-container">
          <unnnic-icon-stub data-v-68ebadeb="" filled="false" next="false" icon="test-icon" clickable="false" size="md" scheme="neutral-darkest" data-testid="title-icon" class="unnnic-modal-dialog__container__title-icon"></unnnic-icon-stub>
          <h1 data-v-68ebadeb="" class="unnnic-modal-dialog__container__title-text" data-testid="title-text">Test Title</h1>
        </section>
        <unnnic-icon-stub data-v-68ebadeb="" filled="false" next="false" icon="close" clickable="true" size="md" scheme="neutral-cloudy" data-testid="close-icon"></unnnic-icon-stub>
      </header>
      <section data-v-68ebadeb="" class="unnnic-modal-dialog__container__content"></section>
      <section data-v-68ebadeb="" data-testid="actions-section" class="unnnic-modal-dialog__container__actions">
        <unnnic-button-stub data-v-68ebadeb="" size="large" text="Cancel" type="tertiary" float="false" iconleft="" iconright="" iconcenter="" iconsfilled="false" next="false" disabled="false" loading="false" data-testid="secondary-button" class="unnnic-modal-dialog__container__actions__secondary-button"></unnnic-button-stub>
        <unnnic-button-stub data-v-68ebadeb="" size="large" text="Confirm" type="primary" float="false" iconleft="" iconright="" iconcenter="" iconsfilled="false" next="false" disabled="false" loading="false" data-testid="primary-button" class="unnnic-modal-dialog__container__actions__primary-button"></unnnic-button-stub>
      </section>
    </section>
  </section>
</section>"
`;
